### ISSUE TITLE
Invalid HTTP message when redirecting POST by 302

### ISSUE THREAD

**Issue by pps83** [curl/curl#2554]
### I did this

I POST data to a server that redirects to another url using 302 status code. When following the redirect libcurl seem to be creating invalid http message:

```
POST /test/a/b/test123 HTTP/1.1
Host: 127.0.0.1:61706
User-Agent: ABC
Accept: */*
Accept-Encoding: deflate, gzip
Referer: http://127.0.0.1:61706/test/a/b/c
Content-Encoding: gzip
```

Note, that when server redirects with 307 then libcurl does proper repost. Is the message quoted above valid? I use my own http server for unit testing and I treat this message as invalid and return 411.

### I expected the following

I expect the POST to be reposted to new url, or fail or do a get. Seems that libcurl reposts it, however the message imo is corrupt since it doesn't contain `Content-Length` and it doesn't have chunked transfer headers.

### curl/libcurl version

#define LIBCURL_VERSION "7.59.0"

### operating system

Win10

**Comment by bagder** [comment]
Please tell us how to reproduce this. Like perhaps a small stand-alone source code.

**Comment by pps83** [comment]
node:
```
var server = require('http').createServer((req,res) =>
{
    console.log('\n\nmethod:', req.method);
    console.log('raw request headers:', req.rawHeaders);
    res.writeHead(302, {"Location" : "/test", "Content-Type" : "text/plain", "Content-Length" : "0"});
    res.end();
});
server.listen(3000);
```
curl cmd line:

> curl -d "param1=value1&param2=value2" -X POST http://localhost:3000/data -i -L --max-redirs 1

Change for 307 and results will be OK.

**Comment by pps83** [comment]
That's exactly what the issue is. If you post using libcurl to some url and it redirects the post somewhere else, then libcurl will send malformed request to redirected url. (as long as the message I quoted in original message is malformed for a POST request).

**Comment by bagder** [comment]
Sorry, I don't understand. Can you show us a full request + response and highlight what the malformed part is?

**Comment by pps83** [comment]
```
POST /test/a/b/test123 HTTP/1.1
Host: 127.0.0.1:61706
User-Agent: ABC
Accept: */*
Accept-Encoding: deflate, gzip
Referer: http://127.0.0.1:61706/test/a/b/c
Content-Encoding: gzip
```

Is this valid POST request or not? AFAIK, post has to have content-length or it has to be chunked transfer, otherwise you don't know where the end of the post message is. Just run that sample nodejs script and use that curl cmd line I showed and nodejs will print all rawHeaders. First request is OK, but on redirected one curl posts without content-length and wihtout using chunked transfer. I though that post requires that there is a content length, or chunked transfer to be able to parse incoming request.

**Comment by bagder** [comment]
[RFC 7230, section 3.3](https://tools.ietf.org/html/rfc7230#section-3.3) says
> The presence of a message body in a request is signaled by a Content-Length or Transfer-Encoding header field.

So without those headers, the request has no body. That seems legal but possibly not what you want...

**Comment by pps83** [comment]
Yes, looks like it might be legal. [However](https://tools.ietf.org/html/rfc7230#section-3.3.2),

>   A user agent SHOULD send a Content-Length in a request message when
   no Transfer-Encoding is sent and the request method defines a meaning
   for an enclosed payload body.  For example, a Content-Length header
   field is normally sent in a POST request even when the value is 0
   (indicating an empty payload body).  A user agent SHOULD NOT send a
   Content-Length header field when the request message does not contain
   a payload body and the method semantics do not anticipate such a
   body.

The other question. Why would it follow POST redirect but without actually posting the body? Is that OK?

**Comment by bagder** [comment]
> Why would it follow POST redirect but without actually posting the body? Is that OK?

That's what the option combination you've used does. If you don't think it's ok, then I don't think you should use that combo.

I don't think we have a way to tell curl to actually POST with the full body again when also told to follow a redirect that changes to GET.

If you really want the body to be sent again when following a 302 redirect, `--post302` is the proper way.

**Comment by pps83** [comment]
> If you really want the body to be sent again in a POST when following a 302 redirect, --post302 is the proper way.

Thanks, I read the docs but didn't think about that option. Just in case, if anybody has similar issue, these are [CURLOPT_POSTREDIR](https://curl.haxx.se/libcurl/c/CURLOPT_POSTREDIR.html) and CURL_REDIR_POST_302 flag.


> I don't think we have a way to tell curl to actually POST with the full body again when also told to follow a redirect that changes to GET.

According to [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Location):

> 301 (Permanent Redirect) and 302 (Found) doesn't change the method most of the time, though older user-agents may (so you basically don't know).

Seems that was old behavior to switch to GET on 302 redirects. Also, why are you talking about GET? [There are no GETs involved](https://github.com/curl/curl/issues/2554#issuecomment-387686289). Second request is also a POST with no body (just like GET, but still uses POST verb).
Is the second request a GET request?

**Comment by bagder** [comment]
First, the canonical spec for HTTP response code 302 is of course [RFC 7231, section 6.4.3](https://tools.ietf.org/html/rfc7231#section-6.4.3).

I'm talking about GET because a 302 changes the method to GET from whatever it was before. That's what all the browsers do for 302 and that's what curl does. The spec says "a user agent MAY change the request method from POST to GET for the subsequent request" but curl mimics the browsers here and provides an option to change the behavior.

**Comment by pps83** [comment]
You omit important part of that spec quote. In full it's:

> For historical reasons, a user agent MAY change the request method from POST to GET for the subsequent request.

For historical reasons user agents MAY do so, it doesn't make it correct. Spec simply allows that old historical behavior. Curl mimics what browsers do? Fine, but that's not the point.
The point is: there is no GET involved at all. isn't it? I get second post request with no body.

I think that body had to be either re-posted, or request had to convert to GET (to mimic historical behavior). What I currently get is this: I get POST with no body, and there is no Content-Length or transfer encoding for that post request without body (which seems to be acceptable according to spec, while curl SHOULD include content-length: 0 for such request). If we assume that this post with no content-length is OK, I don't see how it could possibly be ok to still POST, but post something else?! E.g. I posted body, but destination ended up getting POST with empty body?! That is the point and I don't see how that could be acceptable.

**Comment by bagder** [comment]
> The point is: there is no GET involved at all. isn't it? I get second post request with no body.

302 means changing to GET. You then override that decision with `-X POST` and tells curl you insist on replacing the method keyword with `POST` instead of `GET`. As the documentation for `-X` explains, it doesn't change the underlying behavior, it only changes the actual string used in the request. Thus for this 302 redirect with your choice of command line options you get GET behavior with a POST method.

> I posted body, but destination ended up getting POST with empty body?! That is the point and I don't see how that could be acceptable.

That happened because you told curl you know better how to behave and then when it didn't do what you expected it to, you blame curl for not figuring out what you intended that it should do.

You seem to be fishing for me or someone to say that this is a bug, but I still haven't seen anything here that indicates that it is. You've just used the wrong command line option here to do what you want.
